plugins/cinema4d/import_cast.pyp

```diff
@@ -417,16 +425,82 @@ def importSkeletonNode(modelNull, skeleton):
 
     return boneIndexes
 
-def importAnimationNode(doc, animation, path):
+
+def importAnimationNode():
     gui.MessageDialog(text="Animations are currently not supported.", type=c4d.GEMB_ICONSTOP)
 
-def importInstanceNodes(doc, nodes, path):
-    gui.MessageDialog(text="Instances are currently not supported.", type=c4d.GEMB_ICONSTOP)
+
+def importInstanceNodes(doc, nodes, node, path):
+    rootPath = c4d.storage.LoadDialog(title='Select the root directory where instance scenes are located', flags=2)
+
+    if rootPath is None:
+        return gui.MessageDialog(text="Unable to import instances without a root directory!", type=c4d.GEMB_ICONSTOP)
+
+    uniqueInstances = {}
+
+    for instance in nodes:
+        refs = os.path.join(rootPath, instance.ReferenceFile().Path())
+
+        if refs in uniqueInstances:
+            uniqueInstances[refs].append(instance)
+        else:
+            uniqueInstances[refs] = [instance]
+
+    name = os.path.splitext(os.path.basename(path))[0]
+
+    # Create a collection for our objects
+    rootNull = BaseObject(c4d.Onull)
+    rootNull.SetName(name)
+    doc.InsertObject(rootNull)
+
+    instanceNull = BaseObject(c4d.Onull)
+    instanceNull.SetName("%s_instances" % name)
+    instanceNull.InsertUnder(rootNull)
+
+    sceneNull = BaseObject(c4d.Onull)
+    sceneNull.SetName("%s_scenes" % name)
+    sceneNull.InsertUnder(rootNull)
+    # Disable source models visibility
+    sceneNull[c4d.ID_BASEOBJECT_VISIBILITY_EDITOR] = 1
+    sceneNull[c4d.ID_BASEOBJECT_VISIBILITY_RENDER] = 1
+
+    for instancePath, instances in uniqueInstances.items():
+        instanceName = os.path.splitext(os.path.basename(instancePath))[0]
+
+        try:
+            cast = Cast.load(instancePath)
+            for root in cast.Roots():
+                for child in root.ChildrenOfType(Model):
+                    modelNull = importModelNode(doc, node, child, instancePath)
+            modelNull.InsertUnder(sceneNull)
+        except:
+            print("Failed to import instance: %s" % instancePath)
```

Comment:
Is there a better way to notify the user? Does this popup anywhere? See Maya/Blender reporting.

Reply:
This will print in the Cinema 4D console. I could make a popup saying that something went wrong and to check console for the detailed list of what models failed.